[ZIP 212] Allow recipient to derive Sapling ephemeral secret from note plaintext (#222)
zcash/zips: zip-seanbowe-esktransfer.rst

ZIP: XXX

> Title: Transfer Sapling Ephemeral Secret to Receipient in Note Plaintext

Reviewer: Typo: "Receipient"

Owners: Sean Bowe <sean@z.cash>
Status: Draft
Category: Consensus
Created: 2019-03-31
License: MIT

Terminology
===========

The key words "MUST" and "MUST NOT" in this document are to be interpreted as described in RFC 2119.

Abstract
========

> This ZIP proposes a change to the transaction structure to accomodate a new field in the Note Plaintext transmitted to recipients of Sapling payments, the purpose of which is to avoid an assumption on zk-SNARK knowledge soundness for preventing diversified address linkability.

Reviewer: Typo: "accomodate"

In order to prevent this attack, the protocol currently forces the sender to prove knowledge of the discrete logarithm of ``epk`` with respect to the ``g_d`` included within the Note commitment. This ``g_d`` is determined by ``d`` and recomputed during Note decryption, and so the recipient will either be unable to decrypt the note or the sender will be unable to perform the attack.

However, this check occurs as part of the zero-knowledge proof statement and so relies on the knowledge soundness of the underlying zk-SNARK in Sapling. It would be preferable to force the sender to transfer ``esk`` within the Note plaintext so that, during note decryption, the recipient can ensure that it corresponds with ``epk`` and ``d`` correctly and ignore the payment as invalid otherwise. This allows a line of defense in the case that knowledge soundness of the zk-SNARK does not hold.

Specification
=============

The transaction format is modified (in coordination with other proposals) so that the Sapling note plaintext must contain an additional 32-byte ``esk`` field.

> Sapling note decryption is defined to require that the ``epk = [esk] GH(d)`` for all valid Note plaintexts.

Reviewer 2: Nit: inconsistent capitalisation of "note plaintext".

Rationale
=========

> The attack that this prevents is an interactive attack that requires an adversary to be able to break critical soundness properties of the zk-SNARKs underlying Sapling. It is potentially valid to assume that this cannot occur, due other damaging effects on the system such as undetectable counterfeiting. However, we have attempted to avoid any instance in the protocol where privacy (even against interactive attacks) depended on strong cryptographic assumptions. Acting differently here would be confusing for users that have previously been told that "privacy does not depend on zk-SNARK soundness" or similar claims.

Reviewer 2: s/due other/due to other/

It is possible for us to infringe on the length of the ``memo`` field and ask the sender to provide ``esk`` within the existing Note plaintext without modifying the transaction format, but this would harm users that have come to expect a 512-byte memo field to be available to them. Changes to the memo field length should be considered in a broader context than changes made for cryptographic purposes.

> It is possible to transmit a signature of knowledge of a correct ``esk`` rather than ``esk`` itself, but this appears to be an unnecessary complication and is likely slower than just supplying ``esk``.

Reviewer 2: This will need some more in-depth analysis of whether the assumption made of the encryption and key exchange needs to change. That doesn't block publication as a draft.

Security and Privacy Considerations
===================================

> The changes made in this proposal prevent an interactive attack that could link together diversified addresses by only breaking the knowledge soundness assumption of the zk-SNARK. It is already assumed that the adversary cannot defeat the ECDLP assumption of the Jubjub elliptic curve, for it could perform a linkability attack trivially in that case.

Reviewer 2: I believe it's EC-DDH rather than EC-DLP.

Author: Given `(pk_d, d)` and `(pk_d', d')`, breaking EC-DLP is sufficient to link the two addresses (by extracting the same `ivk`). I don't think EC-DDH applies here.

Reviewer 2: It's definitely EC-DDH. See the non-normative notes in section 5.4.1.6 of the spec.

(The TODO comments there don't affect this point.)

Author: Oh, you're right.

::

  ZIP: 212

>   Title: Transfer Sapling Ephemeral Secret to Recipient in Note Plaintext

Reviewer 2: ```suggestion
  Title: Allow recipient to derive Sapling ephemeral secret from note plaintext
```
(since we're not proposing to literally include the secret in the note plaintext any more).

The feature works by allowing payment addresses to become a tuple (``pk_d``, ``d``) of a public key ``pk_d`` and 88-bit diversifier ``d`` such that ``pk_d = [ivk] GH(d)`` for some incoming viewing key ``ivk``. The hash function ``GH(d)`` maps from a diversifier to prime order elements of the Jubjub elliptic curve. It is possible for a user to choose many ``d`` to create several distinct and unlinkable payment addresses of this form.

In order to make a payment to a Sapling address, an ephemeral secret ``esk`` is sampled by the sender and an ephemeral public key ``epk = [esk] GH(d)`` is included in the Output description. Then, a shared Diffie-Hellman secret is computed by the sender as ``[esk] [8] pk_d``. The recipient can recover this shared secret without knowledge of the particular ``d`` by computing ``[ivk] [8] epk``. This shared secret is then used as part of Note decryption.

> Naively, the recipient cannot know which ``(pk_d, d)`` was used to compute the shared secret, but the sender is asked to include the ``d`` within the note plaintext to reconstruct the Note. However, if the recipient has many known addresses, an attacker could use a different payment address to perform secret exchange and, by observing the behavior of the recipient, link the two diversified addresses together.

Reviewer 3: It would help to spell out this attack, or point to it. My intuitive sense is that the attacker would use a linear combination of known addresses, but this is not made clear.

Author: The trick is for the attacker to send a payment to one of the diversified addresses, but for the key exchange they use another diversified address they suspect the recipient also owns. The recipient would have no idea that this happened, since they cannot know the real `pk_d`/`g_d` that the sender used for the key exchange.

If the recipient then went on to say "Thanks for the payment!" the attacker would know that both addresses are owned by the recipient, linking them.

> In order to prevent this attack, the protocol currently forces the sender to prove knowledge of the discrete logarithm of ``epk`` with respect to the ``g_d`` included within the Note commitment. This ``g_d`` is determined by ``d`` and recomputed during Note decryption, and so the recipient will either be unable to decrypt the note or the sender will be unable to perform the attack.

Reviewer 3: How is ``g_d`` determined exactly? Is it ``g^d`` or something else? This part is not quite self-contained

Author: Thanks! I didn't describe what `g_d` was defined as.

> However, this check occurs as part of the zero-knowledge proof statement and so relies on the knowledge soundness of the underlying zk-SNARK in Sapling. It would be preferable to force the sender to transfer ``esk`` within the Note plaintext so that, during note decryption, the recipient can ensure that it corresponds with ``epk`` and ``d`` correctly and ignore the payment as invalid otherwise. This allows a line of defense in the case that knowledge soundness of the zk-SNARK does not hold.

Reviewer 3: Instead of "corresponds with" it would help to specify exactly what the recipient checks.
